index.js

```diff
@@ -253,12 +229,34 @@
     );
 
     typeClass.methods = {};
-    staticMethodNames.forEach (_name => {
-      typeClass.methods[_name] = getStaticMethod (_name);
+
+    staticMethods.forEach (method => {
+      const _name = method.name;
+      typeClass.methods[_name] = (
+        method.arity === 0 ? typeRep => (
+          staticMethod (_name, typeRep) ()
+        ) :
+        method.arity === 1 ? (typeRep, a) => (
+          staticMethod (_name, typeRep) (a)
+        ) :
+        (typeRep, a, b) => (
+          staticMethod (_name, typeRep) (a, b)
+        )
+      );
     });
-    prototypeMethodNames.forEach (_name => {
-      typeClass.methods[_name] = x => (
-        (getPrototypeMethod (_name) (x)).bind (x)
+
+    prototypeMethods.forEach (method => {
+      const _name = method.name;
+      typeClass.methods[_name] = (
+        method.arity === 0 ? a => (
+          (prototypeMethod (_name, a)).call (a)
+        ) :
+        method.arity === 1 ? (a, b) => (
+          (prototypeMethod (_name, a)).call (a, b)
+        ) :
+        (a, b, c) => (
+          (prototypeMethod (_name, a)).call (a, b, c)
+        )
       );
```

Comment:
This bothers me a little. Is this approach significantly more efficient than `(target, ...args) => ...`?

Reply:
The whole thing is in a separate commit, and I wasn't sure from the start whether the gain are really worth it (https://github.com/sanctuary-js/sanctuary-type-classes/pull/151#issuecomment-824726722). I'll rebase to move this commit to the end, and then run the benchmarks with, and without it to see.

Reply:
It proved non-trivial to reorder the commits, so I manually reverted the arity optimization in a new commit (81442658c3439c1a38862ff5ac3d835f937bce66). I ran the benchmarks against `master` (to compare the results with the previous run against `master`) but I also ran them against `HEAD^` to see the absolute difference. The results were consistent with each other, except for the `methods.sort.Array` suite. Here's the difference compared to `HEAD^` (dd4f87d1bcbdfa194add977973a5cb86d23d305a):

```
┌─────────────────────────────┬─────────────────────────────┬─────────────────────────────┬────────┬─────────┬───┐
│ suite                       │ dd4f87d (HEAD^)             │ 8144265 (HEAD)              │ diff   │ change  │ α │
├─────────────────────────────┼─────────────────────────────┼─────────────────────────────┼────────┼─────────┼───┤
│ functions.empty.Array       │ 90,335,519 Hz ±1.06% (n 92) │ 58,552,936 Hz ±0.54% (n 93) │ 021.3% │ -035.2% │ ✗ │
├─────────────────────────────┼─────────────────────────────┼─────────────────────────────┼────────┼─────────┼───┤
│ functions.of.Array          │ 94,248,155 Hz ±1.10% (n 92) │ 50,486,427 Hz ±0.48% (n 93) │ 030.2% │ -046.4% │ ✗ │
├─────────────────────────────┼─────────────────────────────┼─────────────────────────────┼────────┼─────────┼───┤
│ methods.sort.List           │ 279 Hz ±1.52% (n 87)        │ 284 Hz ±1.61% (n 85)        │ 000.8% │ +001.6% │   │
└─────────────────────────────┴─────────────────────────────┴─────────────────────────────┴────────┴─────────┴───┘
```

And here's the results compared to `master` (c230476e96bfb3c8d24ef0d86b10b1051f2848f3):

```
┌─────────────────────────────┬────────────────────────────┬─────────────────────────────┬────────┬─────────┬───┐
│ suite                       │ c230476 (master)           │ 8144265 (HEAD)              │ diff   │ change  │ α │
├─────────────────────────────┼────────────────────────────┼─────────────────────────────┼────────┼─────────┼───┤
│ functions.empty.Array       │ 6,213,442 Hz ±1.46% (n 92) │ 58,849,793 Hz ±0.63% (n 90) │ 080.9% │ +847.1% │ ✓ │
├─────────────────────────────┼────────────────────────────┼─────────────────────────────┼────────┼─────────┼───┤
│ functions.of.Array          │ 5,935,312 Hz ±1.12% (n 92) │ 53,583,473 Hz ±0.98% (n 92) │ 080.1% │ +802.8% │ ✓ │
├─────────────────────────────┼────────────────────────────┼─────────────────────────────┼────────┼─────────┼───┤
│ methods.sort.List           │ 351 Hz ±1.60% (n 88)       │ 275 Hz ±1.89% (n 85)        │ 012.1% │ -021.6% │ ✗ │
└─────────────────────────────┴────────────────────────────┴─────────────────────────────┴────────┴─────────┴───┘
```

And for reference, `HEAD^` compared to `master`:

```
┌─────────────────────────────┬────────────────────────────┬─────────────────────────────┬────────┬──────────┬───┐
│ suite                       │ c230476 (master)           │ dd4f87d (HEAD^)             │ diff   │ change   │ α │
├─────────────────────────────┼────────────────────────────┼─────────────────────────────┼────────┼──────────┼───┤
│ functions.empty.Array       │ 5,758,854 Hz ±1.30% (n 92) │ 85,464,333 Hz ±1.11% (n 93) │ 087.4% │ +1384.1% │ ✓ │
├─────────────────────────────┼────────────────────────────┼─────────────────────────────┼────────┼──────────┼───┤
│ functions.of.Array          │ 5,688,936 Hz ±0.75% (n 93) │ 90,316,006 Hz ±0.87% (n 95) │ 088.1% │ +1487.6% │ ✓ │
├─────────────────────────────┼────────────────────────────┼─────────────────────────────┼────────┼──────────┼───┤
│ methods.sort.List           │ 349 Hz ±1.65% (n 88)       │ 354 Hz ±0.88% (n 88)        │ 000.7% │ +001.5%  │   │
└─────────────────────────────┴────────────────────────────┴─────────────────────────────┴────────┴──────────┴───┘
```

There's a strange discrepancy (which I'm getting consistently) when testing `methods.sort.List` on HEAD^ against `master`, versus testing it against `HEAD`. When testing against `master`, I'm getting 354 Hz, and when testing the same code against `HEAD` I'm getting 279 Hz. This same discrepancy also exists in the benchmarks I ran a few weeks back (https://github.com/sanctuary-js/sanctuary-type-classes/pull/151#issuecomment-902169382). It could be related to differences in how the heap is filling when so when the garbage collector runs, or something like that. :shrug: 

Either way, we can see performance getting worse for the static methods when the arity optimization is removed. We can also see in the commit itself (81442658c3439c1a38862ff5ac3d835f937bce66) that these static methods no longer report the correct arity. What do you think @davidchambers?

Reply:
It looks like we could keep it around for the static functions, but it's not adding much for the prototype methods.

Reply:
> What do you think @davidchambers?

I defer to you on this, Aldwin. :)